run.c

```diff
@@ -383,35 +400,23 @@ void compute_accelerations(int is_PM, int FirstStep)
     if(FirstStep)
         grav_short_tree();
 
-    if(All.NTotalInit[0] > 0)
+    /* Note this must be after gravaccel and hydro,
+     * because new star particles are not in the tree,
```

Comment:
Can we move PM before all other forces? If so perhaps we can find a way to avoid PM depending on the tree, then defer tree build after PM on a PM step.

Be aware that we currently have to rebuild the tree twice in a PM step. I recall in Gadget-2/3 it was a big fuzz that the tree and PM are overlapped and tree is not built more than once per step. 

I am not sure how much of a gain it really is though.

Reply:
That isn't a big gain, true. But moving PM first might be a good idea anyway. Then the tree builds could all be explicitly in run.c/compute_accelerations() rather than hidden elsewhere.

Reply:
The PM depends on the tree because of the accelerated Region assignment. Does that really help that much, since we have to rebuild the tree to do it?

Reply:
I think it can be modified to depend only on the TopTree (for domain). You can keep a call to build the tree and write a FIXME, till I make the change.

Reply:
Ok, I will do that.

Reply:
Oh, my mistake. We cannot move PM first because the tree opening angle uses GravPM from the last timestep.

Reply:
Uggh. Please add a comment so we don't run into this next time looking at the code! We can then merge the PR.

Reply:
There is a comment! See line 369. I just hadn't read it, and there is no defence against that.